### ISSUE TITLE
Unknown error for send_sms

### ISSUE THREAD

**Issue by stephanfevrier** [Salamek/huawei-lte-api#19]
Hi,

Thank you for developing this api. However, I am getting an error when trying to send an sms using this api:
```
python3 send_sms.py "REDACTED" "Hello World"
{'cdma_enabled': '0', 'url_enabled': '1', 'import_enabled': '1', 'smsisusepdu': '0', 'SendType': '1', 'maxphone': '50', 'UseSReport': '0', 'SaveMode': '0', 'Validity': '10752', 'pagesize': '20', 'smscharlang': '0', 'Sca': '+41787777070'}
Traceback (most recent call last):
  File "send_sms.py", line 16, in <module>
    print(client.sms.send_sms(sys.argv[1], sys.argv[2]))
  File "/home/pi/.local/lib/python3.5/site-packages/huawei_lte_api/AuthorizedConnection.py", line 19, in wrapped
    return fn(*args, **kw)
  File "/home/pi/.local/lib/python3.5/site-packages/huawei_lte_api/api/Sms.py", line 114, in send_sms
    }, dicttoxml_xargs=dicttoxml_xargs)
  File "/home/pi/.local/lib/python3.5/site-packages/huawei_lte_api/Connection.py", line 112, in post
    data = self._check_response_status(self._process_response_xml(response.content))
  File "/home/pi/.local/lib/python3.5/site-packages/huawei_lte_api/Connection.py", line 64, in _check_response_status
    error_code
huawei_lte_api.exceptions.ResponseErrorException: 100005: Unknown
```

My script looks like this:

```
import sys
from huawei_lte_api.Client import Client
from huawei_lte_api.AuthorizedConnection import AuthorizedConnection
from huawei_lte_api.Connection import Connection

connection = AuthorizedConnection('http://admin:REDACTED@192.168.8.1/')

client = Client(connection) # This just simplifies access to separate API group$

print(client.sms.config())
print(client.sms.send_sms(list(sys.argv[1]), sys.argv[2]))
print(client.user.logout())
```

Am I doing anything wrong, or is this a bug? Using version 1.1.3 on Raspbian 9.6.

Thanks in advance for your help.

Stephan

**Comment by Salamek** [comment]
@stephanfevrier Hi,
Check how `list()` works
```python
[sadam@x304 ~]$ python
Python 3.7.2 (default, Dec 29 2018, 21:15:15) 
[GCC 8.2.1 20181127] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> print(list('abc'))
['a', 'b', 'c']
>>> 
```
That is not what you wanted right ? :)

`send_sms` accepts list of phone numbers as a first argument `send_sms(['+420123456789', '+420987654321', 'Hey you two!'])`

Try this code:

```
import sys
from huawei_lte_api.Client import Client
from huawei_lte_api.AuthorizedConnection import AuthorizedConnection
from huawei_lte_api.Connection import Connection

connection = AuthorizedConnection('http://admin:REDACTED@192.168.8.1/')

client = Client(connection) # This just simplifies access to separate API group$

print(client.sms.config())
print(client.sms.send_sms([sys.argv[1]], sys.argv[2]))
print(client.user.logout())
```

**Comment by stephanfevrier** [comment]
I just successfully sent an SMS from my B525-23a SIM Card using the huawei hilink app. So the whole route from external app through API, router, cell provider to my business cell phone seems to work.
Same thing is true using the official huawei web interface.

**Comment by BenjaminKnol** [comment]
hey there @stephanfevrier  I also used this api and got it working perfectly I'll share the code I used to send an SMS:

`client.sms.send_sms([PhoneNumberHere], "Message sent as a string")`

it is imperative you use these data types because otherwise you will just get errors

**Comment by stephanfevrier** [comment]
@BenjaminKnol I tried it again:
```
import sys
from huawei_lte_api.Client import Client
from huawei_lte_api.AuthorizedConnection import AuthorizedConnection
from huawei_lte_api.Connection import Connection

connection = AuthorizedConnection('http://admin:REDACTED@192.168.8.1/')

client = Client(connection) # This just simplifies access to separate API groups, you can use device = Device(connection) if y$

print(client.sms.config())
print(client.sms.send_sms([+41768888888], "Hello World"))
print(client.user.logout())
```

and got
```
{'maxphone': '50', 'url_enabled': '1', 'smscharlang': '0', 'import_enabled': '1', 'SendType': '1', 'SaveMode': '0', 'pagesize': '20', 'smsisusepdu': '0', 'UseSReport': '0', 'Sca': '+41787777070', 'cdma_enabled': '0', 'Validity': '10752'}
Traceback (most recent call last):
  File "send_sms.py", line 11, in <module>
    client.sms.send_sms([+41768888888], "Hello World")
  File "/home/pi/.local/lib/python3.5/site-packages/huawei_lte_api/AuthorizedConnection.py", line 19, in wrapped
    return fn(*args, **kw)
  File "/home/pi/.local/lib/python3.5/site-packages/huawei_lte_api/api/Sms.py", line 114, in send_sms
    }, dicttoxml_xargs=dicttoxml_xargs)
  File "/home/pi/.local/lib/python3.5/site-packages/huawei_lte_api/Connection.py", line 112, in post
    data = self._check_response_status(self._process_response_xml(response.content))
  File "/home/pi/.local/lib/python3.5/site-packages/huawei_lte_api/Connection.py", line 64, in _check_response_status
    error_code
huawei_lte_api.exceptions.ResponseErrorException: 100005: Unknown
```

What next? Is there a possibility for a debugging mode of this api or are there any logs that could tell me what causes the error.

**Comment by BenjaminKnol** [comment]
@stephanfevrier hmm weird error, there are two things which I'm considering could be causing your error:

1. Authorized connection is not correctly being established, you can test this by printing
`print(client.device.information())`

2.Some error is being returned but the SMS still may be sent, could you check this as well?

**Comment by Salamek** [comment]
I could simply debug and fix this, if i had required HW or remote access to it... so if someone can provide one of these:
* direct remote access to testing HW (VPN or Public IP or SSH to machine in same network)
* Send me testing HW
* Send me $ to buy testing HW
...

**Comment by technarf** [comment]
> @technarf the only thing that I can think of is a problem that I had that I don't precisely know how it got fixed. I have the modem: B525s-23a for which the API was a bit broken when sending SMS. The master got updated with some stuff which I didn't have because I was working with my own fork. If this is the case for you try to update your fork with the newest code in the master

Thanks for your response. I'm not sure to understand what I have to do...
If I download your fork, it would be ok or not ?
I don't know anything in python so please be indulgent.

**Comment by stephanfevrier** [comment]
With my very limited debugging skils, this is what I found out:

```
pi@hassbian:/home/homeassistant/.homeassistant $ sudo python3 send_sms.py +41768888888 "Hello World"
> /home/homeassistant/.homeassistant/send_sms.py(14)<module>()
-> print(client.sms.send_sms([sys.argv[1]], sys.argv[2]))
(Pdb) s
...
> /usr/local/lib/python3.5/dist-packages/huawei_lte_api/Connection.py(105)post()
-> cookies=self.cookie_jar
(Pdb) pp headers
{'Content-Type': 'application/xml',
 '__RequestVerificationToken': 'ZgcfZdMxHtSvZU2Oee0GJ0P0E2v1i0WL'}
(Pdb) pp self.cookie_jar
<RequestsCookieJar[Cookie(version=0, name='SessionID', value='WU0b7OXC0ewScF7BG84dt6MrTxn1KFYxeyaUBGKTzsmeDRlrfr1a8TAGG6Ej35T9s8Uz0RftRY6qEo9yUpR9c99yHw7jBu6zkYhlpj96qcWDDjsKyMGTfItb5Q5nMWgf', port=None, port_specified=False, domain='admin:PASSWORD@192.168.8.1', domain_specified=False, domain_initial_dot=False, path='/', path_specified=True, secure=False, expires=None, discard=True, comment=None, comment_url=None, rest={'HttpOnly': None}, rfc2109=False)]>
(Pdb) s
...
> /usr/lib/python3/dist-packages/requests/api.py(110)post()
-> return request('post', url, data=data, json=json, **kwargs)
(Pdb) pp url
'http://admin:PASSWORD@192.168.8.1/api/sms/send-sms'
(Pdb) pp data
(b'<?xml version="1.0" encoding="UTF-8" ?><request><Date type="str">2019-02-02 '
 b'23:53:24</Date><Phones type="list"><Phone type="str">+41768888888</Phone></P'
 b'hones><Length type="int">11</Length><Sca type="str"></Sca><Index type="int">'
 b'-1</Index><Reserved type="int">1</Reserved><Content type="str">Hello World</'
 b'Content></request>')
(Pdb) pp json
None
(Pdb) s
...
> /usr/lib/python3/dist-packages/requests/api.py(110)post()-><Response [200]>
-> return request('post', url, data=data, json=json, **kwargs)
(Pdb) l
105  	    :param \*\*kwargs: Optional arguments that ``request`` takes.
106  	    :return: :class:`Response <Response>` object
107  	    :rtype: requests.Response
108  	    """
109  	
110  ->	    return request('post', url, data=data, json=json, **kwargs)
111  	
112  	
113  	def put(url, data=None, **kwargs):
114  	    """Sends a PUT request.
115  	
(Pdb) pp url
'http://admin:PASSWORD@192.168.8.1/api/sms/send-sms'
(Pdb) pp data
(b'<?xml version="1.0" encoding="UTF-8" ?><request><Date type="str">2019-02-02 '
 b'23:53:24</Date><Phones type="list"><Phone type="str">+41768888888</Phone></P'
 b'hones><Length type="int">11</Length><Sca type="str"></Sca><Index type="int">'
 b'-1</Index><Reserved type="int">1</Reserved><Content type="str">Hello World</'
 b'Content></request>')
(Pdb) pp json
None
(Pdb) pp kwargs
{'cookies': <RequestsCookieJar[Cookie(version=0, name='SessionID', value='WU0b7OXC0ewScF7BG84dt6MrTxn1KFYxeyaUBGKTzsmeDRlrfr1a8TAGG6Ej35T9s8Uz0RftRY6qEo9yUpR9c99yHw7jBu6zkYhlpj96qcWDDjsKyMGTfItb5Q5nMWgf', port=None, port_specified=False, domain='admin:PASSWORD@192.168.8.1', domain_specified=False, domain_initial_dot=False, path='/', path_specified=True, secure=False, expires=None, discard=True, comment=None, comment_url=None, rest={'HttpOnly': None}, rfc2109=False)]>,
 'headers': {'Content-Type': 'application/xml',
             '__RequestVerificationToken': 'ZgcfZdMxHtSvZU2Oee0GJ0P0E2v1i0WL'}}
(Pdb) s
...
> /usr/local/lib/python3.5/dist-packages/huawei_lte_api/Connection.py(112)post()
-> data = self._check_response_status(self._process_response_xml(response.content))
(Pdb) pp data
{'Content': 'Hello World',
 'Date': '2019-02-02 23:53:24',
 'Index': -1,
 'Length': 11,
 'Phones': ['+41768888888'],
 'Reserved': 1,
 'Sca': ''}
(Pdb) s
...
> /usr/lib/python3/dist-packages/requests/models.py(786)content()->b'<?xml versi...>\n</error>\n'
-> return self._content
(Pdb) pp self._content
(b'<?xml version="1.0" encoding="UTF-8"?>\r\n<error>\n<code>125002</code>\n<mes'
 b'sage/>\n</error>\n')
(Pdb) s
--Call--
> /usr/local/lib/python3.5/dist-packages/huawei_lte_api/Connection.py(35)_process_response_xml()
-> @staticmethod
(Pdb) s
> /usr/local/lib/python3.5/dist-packages/huawei_lte_api/Connection.py(37)_process_response_xml()
-> return xmltodict.parse(xml, dict_constructor=dict)
(Pdb) s
--Call--
> /usr/local/lib/python3.5/dist-packages/xmltodict.py(190)parse()
-> def parse(xml_input, encoding=None, expat=expat, process_namespaces=False,
(Pdb) pp xml_input
(b'<?xml version="1.0" encoding="UTF-8"?>\r\n<error>\n<code>125002</code>\n<mes'
 b'sage/>\n</error>\n')
(Pdb) exit
```

For comparison's sake, I have also recorded what happens when I (successfully) send a SMS through the web interface. The relevant requests were:
```
      {
        "pageref": "page_2",
        "startedDateTime": "2019-01-10T23:16:05.317+01:00",
        "request": {
          "bodySize": 231,
          "method": "POST",
          "url": "http://192.168.8.1/api/sms/send-sms",
          "httpVersion": "HTTP/1.1",
          "headers": [
            {
              "name": "Host",
              "value": "192.168.8.1"
            },
            {
              "name": "User-Agent",
              "value": "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:62.0) Gecko/20100101 Firefox/62.0"
            },
            {
              "name": "Accept",
              "value": "*/*"
            },
            {
              "name": "Accept-Language",
              "value": "en-US,en;q=0.5"
            },
            {
              "name": "Accept-Encoding",
              "value": "gzip, deflate"
            },
            {
              "name": "Referer",
              "value": "http://192.168.8.1/html/smsinbox.html"
            },
            {
              "name": "Content-Type",
              "value": "application/x-www-form-urlencoded; charset=UTF-8"
            },
            {
              "name": "_ResponseSource",
              "value": "Broswer"
            },
            {
              "name": "__RequestVerificationToken",
              "value": "1Zjh0Ve8Vqa6xQBY4eOWczkHJ82lCXfp"
            },
            {
              "name": "X-Requested-With",
              "value": "XMLHttpRequest"
            },
            {
              "name": "Content-Length",
              "value": "231"
            },
            {
              "name": "Cookie",
              "value": "SessionID=hREcxDuQPjHbbvDNrmN38Q9D7WqglacpuH08lZQDWuvNRPCfqCrR0KwZMm3CDYQEZxIqHT9iEULW8fvi44yJUKrUBj0gEv0wBhXYgiJuG8nfganvFmPDdIN50clLo8Fc"
            },
            {
              "name": "Connection",
              "value": "keep-alive"
            }
          ],
          "cookies": [
            {
              "name": "SessionID",
              "value": "hREcxDuQPjHbbvDNrmN38Q9D7WqglacpuH08lZQDWuvNRPCfqCrR0KwZMm3CDYQEZxIqHT9iEULW8fvi44yJUKrUBj0gEv0wBhXYgiJuG8nfganvFmPDdIN50clLo8Fc"
            }
          ],
          "queryString": [],
          "headersSize": 649,
          "postData": {
            "mimeType": "application/x-www-form-urlencoded",
            "params": [
              {
                "name": "<?xml version",
                "value": "\"1.0\" encoding"
              }
            ],
            "text": "<?xml version=\"1.0\" encoding=\"UTF-8\"?><request><Index>-1</Index><Phones><Phone>+41768888888</Phone></Phones><Sca></Sca><Content>Hello World</Content><Length>10</Length><Reserved>1</Reserved><Date>2019-01-10 23:16:05</Date></request>"
          }
        },
        "response": {
          "status": 200,
          "statusText": "OK",
          "httpVersion": "HTTP/1.1",
          "headers": [
            {
              "name": "Cache-Control",
              "value": "no-cache, no-store, max-age=0, must-revalidate"
            },
            {
              "name": "Pragma",
              "value": "no-cache"
            },
            {
              "name": "Content-Type",
              "value": "text/html"
            },
            {
              "name": "X-Download-Options",
              "value": "noopen"
            },
            {
              "name": "X-Frame-Options",
              "value": "SAMEORIGIN"
            },
            {
              "name": "X-XSS-Protection",
              "value": "1; mode=block"
            },
            {
              "name": "__RequestVerificationToken",
              "value": "0Lyda1lVgoqo817FR3mrkw1LaSS1cyMT"
            },
            {
              "name": "Connection",
              "value": "Keep-Alive"
            },
            {
              "name": "Content-Length",
              "value": "65"
            }
          ],
          "cookies": [],
          "content": {
            "mimeType": "text/html",
            "size": 65,
            "text": "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\r\n<response>OK</response>\r\n"
          },
          "redirectURL": "",
          "headersSize": 321,
          "bodySize": 386
        },
        "cache": {},
        "timings": {
          "blocked": 0,
          "dns": 0,
          "connect": 0,
          "ssl": 0,
          "send": 0,
          "wait": 34,
          "receive": 0
        },
        "time": 34,
        "_securityState": "insecure",
        "serverIPAddress": "192.168.8.1",
        "connection": "80"
      }
```

@Salamek Can you do something with this? Let me know what else you would need to get to the source of our problem.

**Comment by Salamek** [comment]
@stephanfevrier 
THX! that looks good

Web GUI request body:
```xml
<?xml version=\"1.0\" encoding=\"UTF-8\"?>
<request>
	<Index>-1</Index>
	<Phones>
		<Phone>+41768888888</Phone>
	</Phones>
	<Sca></Sca>
	<Content>Hello World</Content>
	<Length>10</Length>
	<Reserved>1</Reserved>
	<Date>2019-01-10 23:16:05</Date>
</request>
```

API request body:
```xml
<?xml version="1.0" encoding="UTF-8" ?>
<request>
	<Date type="str">2019-02-02 23:53:24</Date>
	<Phones type="list">
		<Phone type="str">+41768888888</Phone>
	</Phones>
	<Length type="int">11</Length>
	<Sca type="str"></Sca>
	<Index type="int">-1</Index>
	<Reserved type="int">1</Reserved>
	<Content type="str">Hello World</Content>
</request>
```

As you can see, parameter `Length` is differend for web GUI vs API when same string is provided, could you check web GUI JS source code for function calculating Length argument ? It looks like it is ignoring spaces or something...

If you dont know how to find it then just download all JS files loaded by send sms page and put it somewhere so i can check it my self :smiley_cat:

**Comment by stephanfevrier** [comment]
@Salamek 
Apologies, I've put you on a wrong track there. Apart from editing PASSWORD and phone number, I did also edit the message content, but forgot about its relation to the Length property. The original API message did match its length.

**Comment by Salamek** [comment]
@stephanfevrier hmm ok, now when i look on it, you are using Python lower than 3.6 right ?
Cos order of items in generated XML is incorrect (Python 3.6+ keeps dict items order but older versions dont)

XML:
```
<?xml version="1.0" encoding="UTF-8" ?>
<request>
	<Date type="str">2019-02-02 23:53:24</Date>
	<Phones type="list">
		<Phone type="str">+41768888888</Phone>
	</Phones>
	<Length type="int">11</Length>
	<Sca type="str"></Sca>
	<Index type="int">-1</Index>
	<Reserved type="int">1</Reserved>
	<Content type="str">Hello World</Content>
</request>
```
Code:
```python
return self._connection.post('sms/send-sms', {
            'Index': sms_index,
            'Phones': phone_numbers,
            'Sca': sca,
            'Content': message,
            'Length': len(message),
            'Reserved': text_mode.value,
            'Date': from_date.strftime("%Y-%m-%d %H:%M:%S")
        }, dicttoxml_xargs=dicttoxml_xargs)
```

And i heard rumors that some Huawei routers are requiring correct order of items in XML...
So if you colud pls test your code in Python 3.6+ ?
And if this will work, i will refactor whole project to use OrderedDict to fix this issue when using pre 3.6 versions of Python.

**Comment by Salamek** [comment]
Fixes for Python 3.5 in branch https://github.com/Salamek/huawei-lte-api/tree/ordered-dict
Can anyone test if this branch fixes issue under python 3.5?

**Comment by stephanfevrier** [comment]
Solved my issue too. Just took a while until I realized that I needed to use the python3.7 command instead of python3.

Thank you Salamek!
